components/icu/src/lib.rs

```diff
@@ -18,83 +18,100 @@
 //!
 //! # Data Management
 //!
-//! Most functionality relies on data which clients have to provide to the APIs.
-//!
-//! `ICU4X` uses the concept of a [`DataProvider`] to separate data from logic.
-//! Data providers come in many different forms; the following providers are provided
-//! by `ICU4X` in separate crates:
-//! * [`BlobDataProvider`]: uses an in-memory serde-serialized blob. This is the most flexible provider, and
-//!   data can be updated at runtime.
-//! * `BakedDataProvider`: a code-generated provider that contains the data directly in Rust code. This is
-//!   the most efficient provider as it's serialization-free, and allows for compile-time optimizations.
-//! * [`FsDataProvider`]: uses a file system tree of Serde files. This is mostly useful for development and
-//!   not recommended in production for performance reasons.
-//! * [`icu_provider_adapters`]: this crate contains APIs to combine providers or
-//!   provide additional functionality such as locale fallback.
+//! Most internationalization algorithms are data-driven based on surveys of locale experts.
+//! ICU4X offers multiple ways to manage locale data: many clients can start by using the
+//! extensive data compiled into the library, while users with additional requirements can
+//! provide data explicitly using [`DataProvider`]s.
 //!
-//! The data that is required by these providers (in `BakedDataProvider`'s case, the provider itself) can be
-//! generated and customized using the [`icu_datagen`] crate.
+//! ## Compiled data
 //!
-//! # Example
+//! Compiled data is exposed through idiomatic Rust constructors like `new` or `try_new`:
 //!
 //! ```
-//! use icu::calendar::DateTime;
-//! use icu::datetime::{options::length, DateTimeFormatter};
+//! use icu::datetime::DateTimeFormatter;
 //! use icu::locid::locale;
-//! use writeable::assert_writeable_eq;
-//!
-//! let options = length::Bag::from_date_time_style(
-//!     length::Date::Long,
-//!     length::Time::Medium,
-//! )
-//! .into();
 //!
 //! let dtf = DateTimeFormatter::try_new(
-//!     &locale!("es").into(),
-//!     options,
+//!     &locale!("es-US").into(),
+//!     Default::default(),
 //! )
-//! .expect("Failed to create DateTimeFormatter instance.");
+//! .expect("compiled data should include 'es-US'");
+//! ```
+//!
+//! Clients using compiled data benefit from simple code and optimal zero-cost data loading. Additionally,
+//! ICU4X's APIs are designed such that dead-code elimination can optimize away unused compiled data.
```

Comment:
nit: perhaps mention that this is per data type not per locale

Reply:
I don't really want to get into defining what a data key is here.